fix(sync): handle errors of deleted sites

nightnei opened the pull request:
## Related issues

Fixes https://github.com/Automattic/dotcom-forge/issues/9374

## Proposed Changes
With this PR we are handling case when a site was connected, then the user deleted it on wordpress.com, so we need to render error and Deleted state in Sync dialog

## Testing Instructions
1. Open Wordpress.com and purchase a site
2. Open Studio and connect it
3. Return to wordpress.com and delete the site purchased at the first step
4. Return to Studio and assert that you see error: <br /><img width="701" alt="Screenshot 2024-11-21 at 20 03 02" src="https://github.com/user-attachments/assets/6d050366-4271-458d-b2b7-3c8a7108cd5a">
5. Click on "Reconnect" button
6. Assert that you see that deleted site with "Deleted" state:<br ><img width="532" alt="Screenshot 2024-11-21 at 20 03 42" src="https://github.com/user-attachments/assets/c8645900-7b0f-4d9d-9331-8d12e2c9657b">
7. Select some another site in Sync dialog and click "Connect"
8. Assert that new site is connected; Deleted site disappeared; And if you open Sync dialog, you don't see that deleted site anymore

nightnei committed:
fix(sync): handle errors of deleted sites

wojtekn commented on src/components/sync-sites-modal-selector.tsx:
```diff
@@ -129,8 +129,9 @@ const getSortedSites = ( sites: SyncSite[] ) => {
 		'already-connected': 2,
-		'needs-transfer': 3,
-		unsupported: 4,
+		deleted: 3,
```
Could we move `deleted` below `needs-transfer`? The reason is that the latter still have a chance to change their status, but the former can't.

nightnei replied:
Just to share my thoughts - I was thinking about it and it was intentional, exactly there - we render only "deleted && connected", so we will render literally maximum 1 deleted site, and we need to render it at the top, to make it noticeable for user.

But I am ok to move it below `need-transfer` if my argument doesn't sound rational.

wojtekn replied:
It makes sense, thanks for providing more details.

wojtekn commented on src/hooks/use-fetch-wpcom-sites.tsx:
```diff
@@ -106,10 +116,10 @@ export const useFetchWpComSites = ( connectedSiteIds: number[] ) => {
 					filter: 'atomic,wpcom',
 					options: 'created_at,wpcom_staging_blog_ids',
-					site_visibility: 'visible',
+					// site_visibility: 'visible',
```
Can we remove it if we don't need it?

wojtekn reviewed (commented):
I tested it but couldn't make it work. My deleted site still appears as a connected site:

![Screenshot 2024-11-22 at 09 45 23](https://github.com/user-attachments/assets/33763b20-760d-42ab-8897-06d293b46e64)

I tried switching different tabs, restarted Studio twice. Also, when I open 'Connect site' I can't see the delete site there. However, I think this should be expected behavior, as we don't want to display all delete user's sites there.

nightnei commented:
@wojtekn I tried one more time, everything works good, hm....
What I did exactly - opened wordpress.com/me/purchases, deleted `domain` and `plan`, so there are no something connected to my site anymore
Opened `Settings` tab in left sidebar of Calypso, at the bottom deleted site


> Also, when I open 'Connect site' I can't see the delete site there. However, I think this should be expected behavior, as we don't want to display all delete user's sites there.

I was thinking a lot about it, and it was intentional - I render there only `deleted && connected` sites, no more. Why it's important: as @matt-west, mentioned [here](https://github.com/Automattic/dotcom-forge/issues/9374#issuecomment-2478596731) - `If the connection broke because the site was deleted/downgraded it won’t be selectable in the connect dialog, in this case the user can click the Disconnect button to remove it.`, and it totally makes sense, since user opens Sync tab, sees that something wrong, clicks on "Reconnect button" - see there error (e.g. removed / need plan upgrade ) and then decides to - connect another site or close the dialog and disconnect. If we don't show "Removed" state of the domain in Sync dialog, then they wouldn't not what is wrong with their site.

wojtekn commented:
> What I did exactly - opened wordpress.com/me/purchases, deleted domain and plan, so there are no something connected to my site anymore
> Opened Settings tab in left sidebar of Calypso, at the bottom deleted site

In my case, I reverted and removed the site using the Bulk Delete tool. It shows as deleted, see the site 239261314. I've checked the response of `/rest/v1.2/me/sites` Studio receives and site is not there.

> I was thinking a lot about it, and it was intentional - I render there only deleted && connected sites, no more. [...]

Makes sense, thanks for explanation.

wojtekn commented:
In addition to the previous comment, should we treat site as deleted when it doesn't appear in list of returned sites anymore?

nightnei committed:
fix(sync): support Bulk Delete tool, undefined sites in response, updating all connected sites at once and a bit refactoring

nightnei commented:
> In addition to the previous comment, should we treat site as deleted when it doesn't appear in list of returned sites anymore?

Initially I didn't want to do it intentionally, to avoid case when some site is considered deleted by some mistakes, so I wanted to rely only on specific property "is_deleted". But it makes sense, so I was thinking about different edge cases, checked everything carefully and now if there is no site in response, then it has "deleted" state.

nightnei committed:
chore: resolve conflicts

nightnei committed:
chore: resolve conflicts

nightnei committed:
chore: resolve conflicts

nightnei committed:
chore: resolve conflicts

nightnei committed:
test: fix

wojtekn reviewed (approved):
It works as expected now - when I launched Studio, deleted site was marked in the UI correctly as deleted site. Then, I was able to reconnect or disconnect.

nightnei committed:
chore: resolve conflicts
